<a href="https://colab.research.google.com/github/jordano-crt/TelecomX-Challenge/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

Comenzamos, cargando los datos de los links de github provisto para manipulación

In [401]:
import pandas as pd
import requests
import json

In [402]:
URL = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'


In [403]:
response = requests.get(URL)
datos = response.json()
dataFrame_datos = pd.DataFrame(datos)

#🔧 Transformación

### Conociendo la estructura del dataset

El conjunto de datos esta subido a un archivo de tipo json, los cuales son un conjunto de diccionarios, listas o arrays, los cuales pueden estar anidados entres los datos. Como se muestra a continuación.

In [404]:
dataFrame_datos.sample(5)# se debe normalizar

customerID Churn                                           customer  \
5209  7110-BDTWG    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
6104  8345-MVDYC    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
5680  7762-URZQH   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
1925  2722-JMONI   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
4873  6663-JOCQO    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
5209  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
6104     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5680    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
1925     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4873    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                               internet  \
5209  {'InternetService': 'DSL', 'OnlineSecurity': '...   
6104  {'InternetService': 'DSL', 'OnlineSecurity': '...   
5680  {'InternetService': 'Fiber optic', 'OnlineSecu...   
1925  {'InternetService': 'DSL', 'OnlineSecurity': '...   
4873  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
5209  {'Contract': 'Two year', 'PaperlessBilling': '...  
6104  {'Contract': 'Two year', 'PaperlessBilling': '...  
5680  {'Contract': 'Two year', 'PaperlessBilling': '...  
1925  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4873  {'Contract': 'One year', 'PaperlessBilling': '...

Como se observa en *dataFrame_datos.sample(5)*, no estrega 5 ejemplos aleatorios de DataFrame de pandas. Se vizualizan 6 columnas de las cuales 4 desde 'customer' a 'account' son objetos de tipo diccionario, los cuales debemos descomponer es culumnas separadas. Para ello se utilizar *pd.json_normalize()*

In [405]:
datos_nomalizados = pd.json_normalize(datos)

Podemos visualizar la cantidad de filas y columnas que contiene nuestro conjunto de datos con *datos_normalizados.shape*. Que nos muestra un total de 7267 filas con 21 columnas.

In [406]:
datos_nomalizados.shape

(7267, 21)

In [407]:
datos_nomalizados.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
2805  3890-RTCMS    No            Male                       0   
6252  8606-CIQUL    No            Male                       1   
3680  5071-FBJFS    No          Female                       0   
3957  5402-HTOTQ    No            Male                       0   
3863  5294-CDGWY   Yes            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
2805               No                  No               16                Yes   
6252              Yes                 Yes               72                Yes   
3680              Yes                 Yes                4                Yes   
3957               No                  No               16                Yes   
3863              Yes                 Yes               51                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
2805                  No                      DSL  ...                    No   
6252                 Yes              Fiber optic  ...                   Yes   
3680                  No                      DSL  ...                    No   
3957                 Yes                      DSL  ...                   Yes   
3863    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
2805                        No                  Yes                   No   
6252                       Yes                  Yes                  Yes   
3680                        No                   No                   No   
3957                        No                   No                   No   
3863                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
2805                       No   Month-to-month                      Yes   
6252                      Yes         Two year                      Yes   
3680                       No   Month-to-month                       No   
3957                       No   Month-to-month                       No   
3863                      Yes         One year                       No   

        account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
2805             Mailed check                   49.45                    799  
6252  Credit card (automatic)                  115.75                8399.15  
3680         Electronic check                   50.30                  217.1  
3957         Electronic check                   55.30                 875.35  
3863         Electronic check                   59.30                3014.65  

[5 rows x 21 columns]

Gracias a la normalización, se puede apreciar todas las columnas que esta relacionadas con los clientes, de forma más comprensible. Ahora se mostrara una descripcion del DataFrame con *datos_nomalizados.info()*.

In [408]:
datos_nomalizados.info()# como son los datos en el DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

### Limpieza de datos

Se observa que **account.Charges.Total** esta en otro formato, asi que lo debemos cambiar el tipo de dato a float.

In [409]:
# Corrige espacios en blanco y convierte a float, reemplazando errores con NaN
datos_nomalizados['account.Charges.Total'] = pd.to_numeric(
    datos_nomalizados['account.Charges.Total'], errors='coerce'
)

In [410]:
datos_nomalizados.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

Antes de proseguir se debe verificar si hay valores nulos dentro de los datos.O caracteres que no debirian de estar.

In [411]:
datos_nomalizados['account.Charges.Total'].isnull().sum()

np.int64(11)

In [412]:
for datos in datos_nomalizados.columns:
    print(datos, datos_nomalizados[datos].isnull().sum())
print(datos_nomalizados.shape)

customerID 0
Churn 0
customer.gender 0
customer.SeniorCitizen 0
customer.Partner 0
customer.Dependents 0
customer.tenure 0
phone.PhoneService 0
phone.MultipleLines 0
internet.InternetService 0
internet.OnlineSecurity 0
internet.OnlineBackup 0
internet.DeviceProtection 0
internet.TechSupport 0
internet.StreamingTV 0
internet.StreamingMovies 0
account.Contract 0
account.PaperlessBilling 0
account.PaymentMethod 0
account.Charges.Monthly 0
account.Charges.Total 11
(7267, 21)


In [413]:
for columna in datos_nomalizados.columns:
    print(f"\n🔹 Columna: {columna}")
    print(f"   🔸 Cantidad de valores únicos: {datos_nomalizados[columna].nunique()}")
    print(f"   🔸 Valores únicos: {datos_nomalizados[columna].unique()[:10]}")
    #if datos_nomalizados[columna].nunique() > 10:
    #    print("   ⚠️  (Mostrando solo los 10 primeros)")



🔹 Columna: customerID
   🔸 Cantidad de valores únicos: 7267
   🔸 Valores únicos: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' '0011-IGKFF' '0013-EXCHZ'
 '0013-MHZWF' '0013-SMEOE' '0014-BMAQU' '0015-UOCOJ' '0016-QLJIS']

🔹 Columna: Churn
   🔸 Cantidad de valores únicos: 3
   🔸 Valores únicos: ['No' 'Yes' '']

🔹 Columna: customer.gender
   🔸 Cantidad de valores únicos: 2
   🔸 Valores únicos: ['Female' 'Male']

🔹 Columna: customer.SeniorCitizen
   🔸 Cantidad de valores únicos: 2
   🔸 Valores únicos: [0 1]

🔹 Columna: customer.Partner
   🔸 Cantidad de valores únicos: 2
   🔸 Valores únicos: ['Yes' 'No']

🔹 Columna: customer.Dependents
   🔸 Cantidad de valores únicos: 2
   🔸 Valores únicos: ['Yes' 'No']

🔹 Columna: customer.tenure
   🔸 Cantidad de valores únicos: 73
   🔸 Valores únicos: [ 9  4 13  3 71 63  7 65 54 72]

🔹 Columna: phone.PhoneService
   🔸 Cantidad de valores únicos: 2
   🔸 Valores únicos: ['Yes' 'No']

🔹 Columna: phone.MultipleLines
   🔸 Cantidad de valores únicos: 3
   🔸 Valores 

Como se puede apreciar, siguiendo el orden de las columnas del dataFrame, la segunda corresponde a 'churn', que son los datos de las personas que ya no estan en la empresa, y resulta que hay otra opcion que es un espacio vacio, que es distinto de null. Ademas de la columna 'account.Charges.Total', que tiene valores nulos.

In [414]:
datos_nomalizados[datos_nomalizados['Churn'] == ''].shape

(224, 21)

In [415]:
datos_nomalizados[datos_nomalizados['Churn'] == ''].sample(3) # ejemplo de 3 datos que contienen valora '' en churn

customerID Churn customer.gender  customer.SeniorCitizen  \
2138  2991-OKAXW                  Male                       0   
2973  4123-VPQTQ                Female                       0   
3827  5235-XATIX                  Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
2138               No                  No               44                 No   
2973               No                  No                3                Yes   
3827               No                 Yes               22                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
2138    No phone service                      DSL  ...                   Yes   
2973                  No                       No  ...   No internet service   
3827                  No              Fiber optic  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
2138                       Yes                   No                  Yes   
2973       No internet service  No internet service  No internet service   
3827                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
2138                      Yes         One year                      Yes   
2973      No internet service   Month-to-month                       No   
3827                       No   Month-to-month                      Yes   

     account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
2138          Mailed check                   53.95                2375.40  
2973      Electronic check                   19.55                  61.05  
3827      Electronic check                   89.40                2001.50  

[3 rows x 21 columns]

In [416]:
datos_nomalizados[datos_nomalizados['account.Charges.Total'].isnull()].index # Indice de los datos que tiene valor nulo

Index([975, 1775, 1955, 2075, 2232, 2308, 2930, 3134, 3203, 4169, 5599], dtype='int64')

In [417]:
datos_nomalizados = datos_nomalizados[datos_nomalizados['Churn'] != ''] # dejame solo los datos que sean distinto de ''

In [418]:
datos_nomalizados[datos_nomalizados['Churn'] == ''].shape

(0, 21)

In [419]:
datos_nomalizados.shape

(7043, 21)

In [420]:
datos_nomalizados = datos_nomalizados[~datos_nomalizados['account.Charges.Total'].isnull()]


In [421]:
datos_nomalizados['account.Charges.Total'].isnull().sum()

np.int64(0)

In [422]:
datos_nomalizados.shape

(7032, 21)

### Columna de cuentas diarias

Utiliza la facturación mensual para calcular el valor diario, proporcionando una visión más detallada del comportamiento de los clientes a lo largo del tiempo.

In [423]:
# Crear la columna 'Cuentas_Diarias' a partir de la facturación mensual
datos_nomalizados['Cuentas_Diarias'] = datos_nomalizados['account.Charges.Monthly'] / 30


In [424]:
datos_nomalizados[['account.Charges.Monthly', 'Cuentas_Diarias']].sample(5)

account.Charges.Monthly  Cuentas_Diarias
386                     54.40         1.813333
4571                    54.75         1.825000
1782                    93.70         3.123333
7121                    20.00         0.666667
7052                    85.85         2.861667

### Estandarizacion y transfomación de datos

La información se estandarizara para una mayor cosistencia, comprensibilidad y adecuada para análisis. Por ejemplo se convertiran valores textuales como 'yes' y 'no' , a valores binario 0 y 1.

In [425]:
datos_nomalizados['Churn'] = datos_nomalizados['Churn'].map({'Yes': 1, 'No': 0})
datos_nomalizados['Churn'].sample(5)

4433    0
26      0
5772    1
2956    0
448     1
Name: Churn, dtype: int64

In [426]:
binarias = [
    'customer.Partner', 'customer.Dependents',
    'phone.PhoneService', 'account.PaperlessBilling'
]

for col in binarias:
    datos_nomalizados[col] = datos_nomalizados[col].map({'Yes': 1, 'No': 0})


In [427]:
tres_opciones_phone = ['phone.MultipleLines']
tres_opciones_internet = [
    'internet.OnlineSecurity', 'internet.OnlineBackup',
    'internet.DeviceProtection', 'internet.TechSupport',
    'internet.StreamingTV', 'internet.StreamingMovies'
]

# Teléfono (no phone service)
for col in tres_opciones_phone:
    datos_nomalizados[col] = datos_nomalizados[col].map({
        'Yes': 1, 'No': 0, 'No phone service': 2
    })

# Internet (no internet service)
for col in tres_opciones_internet:
    datos_nomalizados[col] = datos_nomalizados[col].map({
        'Yes': 1, 'No': 0, 'No internet service': 2
    })


### Columnas relevantes

Luego de corregir ciertos datos, ahora se pasara a escoger los valores, necesarios para llevar a cabo nuestro análisis de evasión. Se seleccionarán únicamente aquellas columnas que aporten información relevante para explicar y predecir el fenómeno de cancelación de clientes (Churn).

Este filtrado de columnas permitirá reducir el ruido en los datos, mejorar la eficiencia del análisis y concentrar el estudio en factores clave como el tipo de contrato, los servicios adicionales contratados, los métodos de pago, la antigüedad del cliente y los patrones de facturación mensual y total.

La selección final de variables se ha realizado considerando tanto el conocimiento del negocio como el análisis exploratorio preliminar, garantizando que las variables elegidas tengan una relación potencial con el comportamiento de abandono de los clientes.


**Churn**: indica si el cliente canceló (1) o permaneció (0).

**customer.tenure**: Meses que el cliente lleva en la empresa; la antigüedad influye en la probabilidad de churn.

**customer.SeniorCitizen**: Segmento etario que puede presentar patrones de retención distintos.

**customer.Partner**: Clientes con pareja suelen mostrar mayor estabilidad contractual.

**customer.Dependents**: Tener dependientes puede indicar mayor necesidad de estabilidad en servicios.

**internet.InternetService**: Tipo de conexión contratada, que influye en costos y satisfacción.

**internet.OnlineSecurity**: Servicio adicional que puede aumentar la permanencia.

**internet.TechSupport**: Acceso a soporte técnico rápido, reduce insatisfacción.

**account.Contract**: Contratos de mayor plazo suelen reducir el churn por penalizaciones o fidelización.

**account.PaperlessBilling**: Facturación electrónica puede estar asociada a diferentes hábitos de pago.

**account.PaymentMethod**: Algunos métodos de pago facilitan la retención (ej. débito automático).

**account.Charges.Monthly**: Monto mensual pagado; valores altos pueden aumentar la probabilidad de churn.

**account.Charges.Total**: Total pagado históricamente; útil para entender antigüedad y valor del cliente.

**Cuentas\_Diarias**: Gasto diario promedio, para un análisis más granular del comportamiento del cliente.


In [428]:
# Lista de variables seleccionadas
columnas_seleccionadas = [
    'Churn',  # Variable objetivo
    'customer.tenure',
    'customer.SeniorCitizen',
    'customer.Partner',
    'customer.Dependents',
    'internet.InternetService',
    'internet.OnlineSecurity',
    'internet.TechSupport',
    'account.Contract',
    'account.PaperlessBilling',
    'account.PaymentMethod',
    'account.Charges.Monthly',
    'account.Charges.Total',
    'Cuentas_Diarias'
]

# Filtrar el DataFrame
datos_selec = datos_nomalizados[columnas_seleccionadas]

# Verificar resultado
datos_selec.sample(5)

Churn  customer.tenure  customer.SeniorCitizen  customer.Partner  \
4326      1               19                       0                 1   
3175      1                2                       0                 0   
1703      0               56                       0                 1   
1806      1                7                       0                 1   
1940      0               41                       0                 0   

      customer.Dependents internet.InternetService  internet.OnlineSecurity  \
4326                    1              Fiber optic                        0   
3175                    0                      DSL                        0   
1703                    1                       No                        2   
1806                    1              Fiber optic                        0   
1940                    0                      DSL                        0   

      internet.TechSupport account.Contract  account.PaperlessBilling  \
4326                     0   Month-to-month                         1   
3175                     0   Month-to-month                         0   
1703                     2         Two year                         1   
1806                     0   Month-to-month                         1   
1940                     0         One year                         1   

        account.PaymentMethod  account.Charges.Monthly  account.Charges.Total  \
4326         Electronic check                    83.65                1465.75   
3175             Mailed check                    44.65                  74.90   
1703  Credit card (automatic)                    19.80                1119.90   
1806         Electronic check                    69.00                 506.90   
1940  Credit card (automatic)                    56.30                2364.00   

      Cuentas_Diarias  
4326         2.788333  
3175         1.488333  
1703         0.660000  
1806         2.300000  
1940         1.876667

#📊 Carga y análisis

Para carga y análisis se usaran, las bibliotecas matplotlib y seaborn.

In [429]:
import matplotlib.pyplot as plt
import seaborn as sns

In [430]:
datos_selec.describe()

Churn  customer.tenure  customer.SeniorCitizen  customer.Partner  \
count  7032.000000      7032.000000             7032.000000       7032.000000   
mean      0.265785        32.421786                0.162400          0.482509   
std       0.441782        24.545260                0.368844          0.499729   
min       0.000000         1.000000                0.000000          0.000000   
25%       0.000000         9.000000                0.000000          0.000000   
50%       0.000000        29.000000                0.000000          0.000000   
75%       1.000000        55.000000                0.000000          1.000000   
max       1.000000        72.000000                1.000000          1.000000   

       customer.Dependents  internet.OnlineSecurity  internet.TechSupport  \
count          7032.000000              7032.000000           7032.000000   
mean              0.298493                 0.718857              0.722412   
std               0.457629                 0.796556              0.795571   
min               0.000000                 0.000000              0.000000   
25%               0.000000                 0.000000              0.000000   
50%               0.000000                 1.000000              1.000000   
75%               1.000000                 1.000000              1.000000   
max               1.000000                 2.000000              2.000000   

       account.PaperlessBilling  account.Charges.Monthly  \
count               7032.000000              7032.000000   
mean                   0.592719                64.798208   
std                    0.491363                30.085974   
min                    0.000000                18.250000   
25%                    0.000000                35.587500   
50%                    1.000000                70.350000   
75%                    1.000000                89.862500   
max                    1.000000               118.750000   

       account.Charges.Total  Cuentas_Diarias  
count            7032.000000      7032.000000  
mean             2283.300441         2.159940  
std              2266.771362         1.002866  
min                18.800000         0.608333  
25%               401.450000         1.186250  
50%              1397.475000         2.345000  
75%              3794.737500         2.995417  
max              8684.800000         3.958333

### Análisis de la tabla datos_selec

El conjunto de datos final contiene 7.032 registros de clientes con las variables seleccionadas para el análisis de evasión (Churn).

Churn: el 26,6% de los clientes han cancelado su servicio, lo que indica un problema de retención importante.<br>

Antigüedad (tenure): los clientes tienen entre 1 y 72 meses de contrato, con una mediana de 29 meses. Esto sugiere que la mitad de la base lleva menos de 2 años y medio en la compañía.<br>

SeniorCitizen: solo el 16% de los clientes tienen 65 años o más.<br>

Partner y Dependents: el 48% tiene pareja y el 29% tiene dependientes, lo que puede influir en la estabilidad contractual.<br>

Servicios adicionales: la mayoría de clientes no cuenta con servicios como seguridad en línea o soporte técnico, o bien no aplican porque no tienen servicio de internet.<br>

Facturación: Cargos mensuales: promedio de 64,8 con un rango de 18,25 a 118,75.<br>

Cargos totales: promedio de 2.283,3 pero con gran variabilidad, dado que dependen del tiempo que el cliente lleva en la empresa.<br>

Cuentas diarias (columna derivada): gasto promedio de 2,16 por día.<br>

Este análisis preliminar permite identificar patrones iniciales y preparar la base para el análisis exploratorio más profundo.<br>

### distribución de evación

#📄Informe final